metis-harvesting/src/main/java/eu/europeana/metis/harvesting/oaipmh/OaiHarvestingIterator.java

```diff
@@ -0,0 +1,218 @@
+          .listIdentifiers(parameters);
+    } catch (RuntimeException | BadArgumentException e) {
+      try {
+        this.close();
+      } catch (IOException ioException) {
+        LOGGER.info("Could not close connection client.", ioException);
+      }
+      throw new HarvesterException(e.getMessage(), e);
+    }
+    return this.source;
+  }
+
+  private static ListIdentifiersParameters prepareListIdentifiersParameters(OaiHarvest harvest) {
+    ListIdentifiersParameters parameters = ListIdentifiersParameters.request()
+                                                                    .withMetadataPrefix(harvest.getMetadataPrefix());
+    if (harvest.getFrom() != null) {
+      parameters.withFrom(harvest.getFrom());
+    }
+    if (harvest.getUntil() != null) {
+      parameters.withUntil(harvest.getUntil());
+    }
+    if (harvest.getSetSpec() != null) {
+      parameters.withSetSpec(harvest.getSetSpec());
+    }
+    return parameters;
+  }
+
+  @Override
+  public void forEachFiltered(ReportingIteration<R> action,
+      Predicate<OaiRecordHeader> filter) throws HarvesterException {
+    forEachWithPostProcessing(action, postProcessing, filter);
+  }
+
+  private <O> void forEachWithPostProcessing(final ReportingIteration<O> action,
+      final RecordOaiHeaderPostProcessing<O> postProcessing, final Predicate<OaiRecordHeader> filter)
+      throws HarvesterException {
+    final SingleIteration<O> singleIteration = new SingleIteration<>(filter, postProcessing, action, oaiClient, harvest);
+    try {
+      while (getOrCreateSource().hasNext()) {
+        final Header header = Optional.of(getOrCreateSource().next())
+                                      .orElseThrow(() -> new HarvesterException("Unexpected null header."));
+        if (singleIteration.process(header) == IterationResult.TERMINATE) {
+          break;
+        }
+      }
+    } catch (RuntimeException e) {
+      throw new HarvesterException("Problem while iterating through OAI headers.", e);
+    }
+  }
+
+  @Override
+  public void forEachNonDeleted(ReportingIteration<R> action) throws HarvesterException {
+    forEachFiltered(action, Predicate.not(OaiRecordHeader::isDeleted));
+  }
+
+  @Override
+  public Integer countRecords() throws HarvesterException {
+    final Parameters parameters = Parameters.parameters().withVerb(Verb.Type.ListIdentifiers)
+                                            .include(prepareListIdentifiersParameters(harvest));
+    try (final InputStream listIdentifiersResponse = performThrowingFunction(oaiClient,
+        client -> client.execute(parameters))) {
+      return readCompleteListSizeFromXML(listIdentifiersResponse);
+    } catch (OAIRequestException | IOException e) {
+      throw new HarvesterException(String.format(
+          "Problem with counting records for endpoint %1$s because of: %2$s",
+          harvest.getRepositoryUrl(), e.getMessage()), e);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
this method is not needed anymore I see `oaiClient` implements `Closable` interface which implements `AutoClosable`. So my suggestion is to remove it

Reply:
Can you please elaborate?
If I remove it then there is no wai to close the oaiClient..